### PR TITLE
Collection of patches for SV-COMP'17

### PR DESCRIPTION
**Pull request by tautschnig**
These are patches by @peterschrammel (graphml fixes), @martin-cs (floating-point fixes) and myself (further witness fixes and a performance improvement). Likely not to be merged as-is, but rather as separated PRs.

### CODE REVIEW

File: src/ansi-c/library/math.c [hunk 1]
```diff
@@ -22,6 +22,31 @@ inline long double __builtin_fabsl(long double d) { return __CPROVER_fabsl(d); }
...
+
+int __builtin_isless(float f, float g) { return f < g;}
+
+/* FUNCTION: __builtin_islessequal */
+
+int __builtin_islessequal(float f, float g) { return f <= g; }
+
+/* FUNCTION: __builtin_islessgreater */
+
+int __builtin_islessgreater(float f, float g) { return (f < g) || (g > f); }
```
Comment (peterschrammel) [inline src/ansi-c/library/math.c]: @martin-cs, this is the last failing one: `(f < g) || (f > g)`

https://github.com/peterschrammel/cbmc/commit/50ded8889ccc801a47c4f612737c3a8aa06d15ac

File: src/config.inc [hunk 2]
```diff
@@ -2,7 +2,7 @@
 BUILD_ENV = AUTO
 
 # Variables you may want to override
-#CXXFLAGS = -Wall -O0 -g -Werror -Wno-long-long -Wno-sign-compare -Wno-parentheses -Wno-strict-aliasing -pedantic
```
Comment (peterschrammel) [inline src/config.inc]: Ouuups...

**Comment by martin-cs** [comment]
On Sun, 2016-12-18 at 15:40 -0800, Peter Schrammel wrote:
> peterschrammel requested changes on this pull request.
> 
> 
> 
> > +
> +/* FUNCTION: __builtin_isgreaterequal */
> +
> +int __builtin_isgreaterequal(float f, float g) { return f >= g; }
> +
> +/* FUNCTION: __builtin_isless */
> +
> +int __builtin_isless(float f, float g) { return f < g;}
> +
> +/* FUNCTION: __builtin_islessequal */
> +
> +int __builtin_islessequal(float f, float g) { return f <= g; }
> +
> +/* FUNCTION: __builtin_islessgreater */
> +
> +int __builtin_islessgreater(float f, float g) { return (f < g) || (g > f); }
> 
> @martin-cs, this is the last failing one: `(f < g) || (f > g)`

Fixed in my branch.

Cheers,
 - Martin

**Comment by tautschnig** [comment]
@peterschrammel @martin-cs Will refresh from your branches and push some updates to address the polymorphic floating-point functions. Need a few minutes to refresh.

**Comment by tautschnig** [comment]
@peterschrammel @martin-cs Warning: Last night's version of "Support polymorphic floating-point functions ..." had had typos in goto-programs/builtin_functions.cpp. I've now fixes this in a revised version of that patch.

**Comment by peterschrammel** [comment]
@tautschnig, @martin-cs, can we split this into two, clean separate PRs? I can take care of the first 8 commits.

**Comment by martin-cs** [comment]
On Mon, 2016-12-19 at 11:16 -0800, Peter Schrammel wrote:
> @tautschnig, @martin-cs, can we split this into two, clean separate PRs? I can take care of the first 8 commits.

I'll refactor / squash my part but I won't have time to implement it
properly.

Cheers,
 - Martin

**Comment by peterschrammel** [comment]
I've made PRs for the first 7 (#369) and no 8 (#370)

**Comment by tautschnig** [comment]
@martin-cs Could you detail/describe somewhere what further changes over all your existing work are necessary?

**Comment by martin-cs** [comment]
Basically, all of the TODOs:

1. We need round-to-integral and remainder __CPROVER_ functions, with
irept types and encodings.

2. We need to use them rather than the __CPROVER_ish functions.

3. We need a RNA rounding mode.

It's not a lot of work but it's probably @martin-cs work and I'll have
time to do it ... soon?  February?

**Comment by lucasccordeiro** [comment]
@tautschnig: May I ask you which patches have already been merged into CBMC master? It would be good if we could start running CBMC over the current version of the SV-COMP benchmarks.

**Comment by marek-trtik** [comment]
It looks like the branch `tautschnig:sv-comp-2017` itself does not compile:

```
## Entering solvers
make  -C solvers
make[1]: Entering directory '/home/marek/root/cbmc-versions/cbmc-tautschnig/src/solvers'
g++ -c -MMD -MP -std=c++11 -O2 -Wall -Werror -Wno-parentheses -I ..        -I ../../glucose-syrup    -DHAVE_GLUCOSE -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -o sat/satcheck_glucose.o sat/satcheck_glucose.cpp
sat/satcheck_glucose.cpp:21:25: fatal error: core/Solver.h: No such file or directory
compilation terminated.
../common:170: recipe for target 'sat/satcheck_glucose.o' failed
make[1]: *** [sat/satcheck_glucose.o] Error 1
```

**Comment by martin-cs** [comment]
Some of the stuff in the previous set of SV-COMP patches (I think this
PR) were kind of slapped together by me at short notice and really
shouldn't be merged as they were.

**Comment by tautschnig** [comment]
Closing in favour of #1532.

**Pull request closed without merging**